ISICbalance/Services/RequestManager.swift

```diff
@@ -10,167 +10,182 @@ import Foundation
+                    "j_password": user.password,
+                    "_eventId_proceed" : ""
+                ]
+
+                let credentialsUrl = responseShibboleth.response?.url?.absoluteString ?? ""
+                return (credentialsUrl, parameters)
+            }
+            .flatMap(.latest, { (credentialsUrl, parameters) -> SignalProducer<DataResponse<String>, RequestError> in
+                return RequestManager.credentialsRequest(credentialsUrl: credentialsUrl, parameters: parameters)
+            })
+            .flatMap(.latest, { responseCredentials -> SignalProducer<DataResponse<String>, RequestError> in
+                do {
+                    let document: Document = try SwiftSoup.parse(responseCredentials.result.value!)
+                    #warning("TODO - check array size")
+                    let form: Element = try document.select("form").array()[0]
+
+                    //get action value from form to check login process
+                    let action: String = try form.attr("action")
+                    if !action.contains("agata.suz.cvut.cz"){
+                        return SignalProducer.init(error: RequestError.loginFailed)
+                    }
+
+                    let inputs = try form.select("input")
+                    //get RelayState
+                    #warning("TODO - check array size")
+                    let inputName1 = try inputs.array()[0].attr("name")
+                    let inputValue1 = try inputs.array()[0].attr("value")
+                    //get SAMLResponse
+                    let inputName2 = try inputs.array()[1].attr("name")
+                    let inputValue2 = try inputs.array()[1].attr("value")
+
+                    let formParameters = [
+                        inputName1 : inputValue1,
+                        inputName2 : inputValue2
+                    ]
+
+                    return RequestManager.balanceSiteRequest(action: action, formParameters: formParameters)
+                } catch {
+                    return SignalProducer.init(error: RequestError.parseError)
+                }
+            })
+    }
+
+    func parseDocument(dataResponse: DataResponse<String>) -> SignalProducer<Balance,RequestError> {
+        return SignalProducer { [weak self] observer, _ in
+            do {
+                let document: Document = try SwiftSoup.parse(dataResponse.result.value!)
+                #warning("TODO - check array size")
+                let bodyElement: Element = try document.select("body").array()[0]
+                let table: Element = try bodyElement.select("tbody").array()[0]
+                let balanceLine: Element = try table.select("td").array()[4]
+                let lineText = balanceLine.ownText()
+                let lineElements = lineText.split(separator: " ")
+                let balance = lineElements[0]
+                let balanceStruct = Balance(balance: "\(balance) Kč")
+                self?.currentBalance.value = balanceStruct
```

Comment:
Tohle bys neměl potřebovat. Máš nějakej input (response ze serveru), něco s ním uděláš (rozparsuješ) a výsledek tý práce (balance) vrátíš jako value toho signálu. Zapsat z tohodle místa do nějakýho stavu je špatně, to je side effect kterej nikdo nikdy nedohledá.

Reply:
Ok. Takže si mám poslouchat na té akci a odchytávat hodnoty? Já jsem byl zrovna z téhle věci trochu zmatenej ze cvik. V mvvm branchi se v userrepository právě tímto způsobem ukládá hodnota. Tam mi ten případ nepřijde jiný 😳

Reply:
Nově observuji v balanceViewModelu hodnotu signálu. Snad je to správně.

Reply:
Aha, poptám se Dominika, proč to tak má udělané, ale pořád si docela stojím za tím, že to není dobře. Napojení na hodnotu ve VM mi přijde OK.